contracts/simple_dex/lib.rs

```diff
@@ -0,0 +1,472 @@
+        // Will revert if not enough allowance was given to the contract by the caller prior to executing this tx
+        #[ink(message)]
+        pub fn deposit(&mut self, deposits: Vec<(AccountId, Balance)>) -> Result<(), DexError> {
+            let this = self.env().account_id();
+            let caller = self.env().caller();
+
+            // check role, only designated account can add liquidity
+            <Self as AccessControlled>::check_role(
+                self.access_control,
+                caller,
+                Role::LiquidityProvider(this),
+                Self::cross_contract_call_error_handler,
+                Self::access_control_error_handler,
+            )?;
+
+            deposits
+                .into_iter()
+                .try_for_each(|(token_in, amount)| -> Result<(), DexError> {
+                    if !self.tokens.contains(&token_in) {
+                        return Err(DexError::UnsupportedToken(token_in));
+                    }
+
+                    // transfer token_in from the caller to the contract
+                    // will revert if the contract does not have enough allowance from the caller
+                    // in which case the whole tx is reverted
+                    self.transfer_from_tx(token_in, caller, this, amount)??;
+                    Ok(())
+                })?;
+
+            Ok(())
+        }
+
+        #[ink(message)]
+        pub fn withdrawal(
+            &mut self,
+            withdrawals: Vec<(AccountId, Balance)>,
+        ) -> Result<(), DexError> {
+            let this = self.env().account_id();
+            let caller = self.env().caller();
+
+            // check role, only designated account can add liquidity
+            <Self as AccessControlled>::check_role(
+                self.access_control,
+                caller,
+                Role::LiquidityProvider(this),
+                Self::cross_contract_call_error_handler,
+                Self::access_control_error_handler,
+            )?;
+
+            withdrawals.into_iter().try_for_each(
+                |(token_out, amount)| -> Result<(), DexError> {
+                    if !self.tokens.contains(&token_out) {
+                        return Err(DexError::UnsupportedToken(token_out));
+                    }
+
+                    // transfer token_out from the contract to the caller
+                    self.transfer_tx(token_out, caller, amount)??;
+                    Ok(())
+                },
+            )?;
+
+            Ok(())
+        }
+
+        /// Alters the swap_fee parameter
+        ///
+        /// Can only be called by the contract's Admin.
+        #[ink(message)]
+        pub fn set_swap_fee(&mut self, swap_fee: u128) -> Result<(), DexError> {
+            if swap_fee.gt(&100) || swap_fee.lt(&0) {
+                return Err(DexError::WrongParameterValue);
+            }
```

Comment:
(Porting from a commit-level comment):

I don't think it can be `swap_fee.lt(&0)` b/c it's an unsigned integer type - it will always be false.